cupy/random/cupy_distributions.cu

```diff
@@ -0,0 +1,252 @@
+    double Ga, Gb;
+    if ((a <= 1.0) && (b <= 1.0)) {
+        double U, V, X, Y;
+        /* Use Johnk's algorithm */
+        while (1) {
+            U = state->rk_double();
+            V = state->rk_double();
+            X = pow(U, 1.0/a);
+            Y = pow(V, 1.0/b);
+            if ((X + Y) <= 1.0) {
+                if (X +Y > 0) {
+                    return X / (X + Y);
+                } else {
+                    double logX = log(U) / a;
+                    double logY = log(V) / b;
+                    double logM = logX > logY ? logX : logY;
+                    logX -= logM;
+                    logY -= logM;
+                    return exp(logX - log(exp(logX) + exp(logY)));
+                }
+            }
+        }
+    } else {
+        Ga = rk_standard_gamma(state, a);
+        Gb = rk_standard_gamma(state, b);
+        return Ga/(Ga + Gb);
+    }
+}
+
+
+__device__ uint32_t rk_interval_32(rk_state* state, int mx, int mask) {
+    uint32_t sampled = state->rk_int() & mask;
+    while(sampled > mx)  {
+        sampled = state->rk_int() & mask;
+    }
+    return sampled;
+}
+
+__device__ uint64_t rk_interval_64(rk_state* state, uint64_t  mx, uint64_t mask) {
+    uint32_t hi= state->rk_int();
+    uint32_t lo= state->rk_int();
+    uint64_t sampled = (static_cast<uint64_t>(hi) << 32 | lo)  & mask;
+    while(sampled > mx)  {
+        hi= state->rk_int();
+        lo= state->rk_int();
+        sampled = (static_cast<uint64_t>(hi) << 32 | lo) & mask;
+    }
+    return sampled;
+}
+
+
+struct interval_32_functor {
+    template<typename... Args>
+    __device__ uint32_t operator () (Args&&... args) {
+        return rk_interval_32(std::forward<Args>(args)...);
+    }
+};
+
+struct interval_64_functor {
+    template<typename... Args>
+    __device__ uint64_t operator () (Args&&... args) {
+        return rk_interval_64(std::forward<Args>(args)...);
+    }
+};
+
+struct beta_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_beta(std::forward<Args>(args)...);
+    }
+};
+
+// There are several errors when trying to do this a full template
+struct exponential_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_standard_exponential(std::forward<Args>(args)...);
+    }
+};
+
+template<typename F, typename T, typename R, typename... Args>
+__global__ void execute_dist(intptr_t state, intptr_t out, ssize_t size, Args... args) {
+    int id = threadIdx.x + blockIdx.x * blockDim.x;
+    R* out_ptr = reinterpret_cast<R*>(out);
+    if (id < size) {
+        T random(id, state);
+        F func;
+        out_ptr[id] = func(&random, std::forward<Args>(args)...);
+    }
+    return;
+}
+
+template <typename F, typename R>
+struct kernel_launcher {
+    template<typename T, typename... Args>
+    void operator()(int bpg, int tpb, cudaStream_t stream, Args&&... args) { 
```

Comment:
we can't have kernels as function members, so we have to resort to dirty tricks like this ...